java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java

```diff
@@ -0,0 +1,270 @@
+    addPoint(name, value, timestamp, DEFAULT_SOURCE, null);
+  }
+
+  @Override
+  public void send(String name, double value, @Nullable Long timestamp, String source) throws IOException {
+    addPoint(name, value, timestamp, source, null);
+  }
+
+  @Override
+  public void send(String name, double value, String source, @Nullable Map<String, String> pointTags) throws IOException {
+    addPoint(name, value, null, source, pointTags);
+  }
+
+  @Override
+  public void send(String name, double value, @Nullable Long timestamp, String source,
+                   @Nullable Map<String, String> pointTags) throws IOException {
+    addPoint(name, value, timestamp, source, pointTags);
+  }
+
+  private void addPoint(@NotNull String name, double value, @Nullable Long timestamp, @NotNull String source,
+                        @Nullable Map<String, String> pointTags) throws IOException {
+
+    //TODO: extra validation required?
+    String point = pointToString(name, value, timestamp, source, pointTags);
+    if (point != null) {
+      synchronized(pointsMutex) {
+        this.buffer.add(point);
+      }
+    }
+  }
+
+  private static String escapeQuotes(String raw) {
+    return StringUtils.replace(raw, quote, escapedQuote);
+  }
+
+  @Nullable
+  static String pointToString(String name, double value, @Nullable Long timestamp, String source,
+                               @Nullable Map<String, String> pointTags) {
+
+    if (StringUtils.isBlank(name) || StringUtils.isBlank(source)) {
+      LOGGER.debug("Invalid point: Empty name/source");
+      return null;
+    }
+
+    StringBuilder sb = new StringBuilder(quote).
+        append(escapeQuotes(name)).append(quote).append(" ").
+        append(Double.toString(value)).append(" ");
+    if (timestamp != null) {
+      sb.append(Long.toString(timestamp)).append(" ");
+    }
+    sb.append("source=").append(quote).append(escapeQuotes(source)).append(quote);
+
+    if (pointTags != null) {
+      for (Map.Entry<String, String> entry : pointTags.entrySet()) {
+        sb.append(' ').append(quote).append(escapeQuotes(entry.getKey())).append(quote).
+            append("=").
+            append(quote).append(escapeQuotes(entry.getValue())).append(quote);
+      }
+    }
+    return sb.toString();
+  }
+
+  @Override
+  public void flush() throws IOException {
+    internalFlush();
+  }
+
+  private void internalFlush() throws IOException {
+
+    if (!isConnected()) {
+        return;
+    }
+
+    List<String> points = getPointsBatch();
+    if (points.isEmpty()) {
+      return;
+    }
+
+    Response response = null;
+    try (InputStream is = pointsToStream(points)) {
+      response = directService.report("graphite_v2", is);
+      if (response.getStatusInfo().getFamily() == Response.Status.Family.SERVER_ERROR ||
+          response.getStatusInfo().getFamily() == Response.Status.Family.CLIENT_ERROR) {
+        LOGGER.debug("Error reporting points, respStatus=" + response.getStatus());
```

Comment:
Probably should look at putting the points back into the buffer (while respecting the maximum size of the buffer)